### ISSUE TITLE
Workbench fails when the Domain Access module is installed

### ISSUE THREAD

**Issue by g011um** [mjordan/islandora_workbench#796]
The [Domain module](https://www.drupal.org/project/domain) allows a Drupal site to serve content for multiple configured domains from the same Drupal instance; essentially a type of multi-site Drupal install.  The Domain Access module allows individual nodes to be tagged as belonging to one of more configured domains.  When installed, the Domain Access module adds a required entity reference field to all content types -- `field_domain_access` which is a reference to a `Domain` entity.

Workbench does not understand Domain entity references, and will throw an error when attempting a `create` or `update` task with records that contain data for this required field.  The error occurs in `workbench_fields.py` in the `create` and `update` methods of the `EntityReferenceField` class.  The variable `target_type` is set to the type of referenced entity.  When the referenced entity type is unknown to workbench, this variable will have an empty/null value and will cause a "UnboundLocalError: local variable 'target_type' referenced before assignment" later in the method.

I have found a fix for this issue which simply requires the following 2 lines to be added to the `create` and `update` methods in `EntityReferenceField` to initialize the `target_type` variable to `domain` when a `Domain` entity is encountered.

```
        if field_definitions[field_name]['target_type'] == 'domain':
            target_type = 'domain'
```

I can submit a pull request for this is that's preferrable

**Comment by mjordan** [comment]
Thanks @g011um I'll run tests on this branch and if nothing breaks, I'll merge. We'll want to document this new feature as well, probably by adding a new section to https://mjordan.github.io/islandora_workbench_docs/fields/#drupal-field-types.

Question: is the fieldname `field_domain_access` hard coded in (or a hard expectation of) the Domain module?

**Comment by g011um** [comment]
@mjordan `field_domain_access` is hard coded into the Domain Access module.  When the module is installed, it is added as a required field to all content types on the system.

**Comment by dara2** [comment]
Hi @mjordan - we have a client using Domain modules and I am running into this issue as well. It passes --check but when I try to run the create task, I get:
```
Born-Digitals-MacBook-Air-4:islandora_workbench Dara$ ./workbench --config create_sample_staging.yml
OK, connection to Drupal at https://i2-staging.digital.library.pitt.edu verified.
"Create" task started using config file create_sample_staging.yml.
Only node IDs for parents created during this session will be used (not using the CSV ID to node ID map).
╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│                                                                              │
│ /Users/Dara_1/islandora_workbench/./workbench:3569 in <module>               │
│                                                                              │
│   3566                                                                       │
│   3567 try:                                                                  │
│   3568 │   if config["task"] == "create":                                    │
│ ❱ 3569 │   │   create()                                                      │
│   3570 │   if config["task"] == "update":                                    │
│   3571 │   │   update()                                                      │
│   3572 │   if config["task"] == "delete":                                    │
│ /Users/Dara_1/islandora_workbench/./workbench:283 in create                  │
│                                                                              │
│    280 │   │   │   # Entity reference fields (taxonomy_term and node).       │
│    281 │   │   │   if field_definitions[custom_field]["field_type"] == "enti │
│    282 │   │   │   │   entity_reference_field = workbench_fields.EntityRefer │
│ ❱  283 │   │   │   │   node = entity_reference_field.create(                 │
│    284 │   │   │   │   │   config, field_definitions, node, row, custom_fiel │
│    285 │   │   │   │   )                                                     │
│    286                                                                       │
│                                                                              │
│ /Users/Dara_1/islandora_workbench/workbench_fields.py:831 in create          │
│                                                                              │
│    828 │   │   subvalues = self.dedupe_values(subvalues)                     │
│    829 │   │   for subvalue in subvalues:                                    │
│    830 │   │   │   subvalue = str(subvalue)                                  │
│ ❱  831 │   │   │   field_values.append({"target_id": subvalue, "target_type" │
│    832 │   │                                                                 │
│    833 │   │   cardinality = int(field_definitions[field_name].get("cardinal │
│    834 │   │   if -1 < cardinality < len(field_values):                      │
╰──────────────────────────────────────────────────────────────────────────────╯
UnboundLocalError: local variable 'target_type' referenced before assignment
```
Is there a PR I could help test?

**Comment by mjordan** [comment]
Yes, the one provided by @g011um , linked above.

**Comment by dara2** [comment]
I didn't see a PR - just the commit and you asking them if they were going to make a PR? I just tried to make one but I'm not a collaborator :) 
Sorry if I am missing something...

**Comment by mjordan** [comment]
Sorry, over at https://github.com/usask-library/islandora_workbench/commit/137971b2629d3613c28e761203b4528f073cf566 I ask if there is a PR forthcoming... since that branch is now 64 commits behind main I'll add these changes to a new branch in https://github.com/mjordan/islandora_workbench this evening and then ask you to test that. That OK?

**Comment by dara2** [comment]
That's awesome - thanks!

**Comment by dara2** [comment]
By the way - I just manually made those edits to my local workbench_fields.py file and it seems to be working!

**Comment by mjordan** [comment]
That is awesome! Thanks for testing and thanks @g011um for the code contribution.

I'd love to get some basic docs to explain under what situation you'd use this (I know absolutely nothing about domain access) and some same CSV to include in the docs. Can one of you provide that?

**Comment by dara2** [comment]
Sure! It may not be til next week but I can definitely provide those.

**Comment by mjordan** [comment]
No problem at all.

**Comment by mjordan** [comment]
Thanks!

**Comment by mjordan** [comment]
I've applied @g011um's fix to the issue_796 branch, which is ready for testing.

**Comment by dara2** [comment]
A little info on how this is used:
When you install the Domain suite of modules, it includes the Domain Access module. This module adds a required field, field_domain_access, which determines on which domain(s) the node/object will be displayed. This field is an entity reference to all domains in the Domain configuration. So in this case, you'll need to add a column to your Workbench CSV with the heading `field_domain_access`. Its values will be the domains in your multi-domain setup, with `_` instead of `.` So for example `example.one.edu` domain would be referenced in your CSV as `example_one_edu`. This is a multivalue field, so you can have more than one domain referenced in the field_domain_access column.
Here's a simple CSV for testing - it assumes you have two domains configured: example.one.edu and example.two.edu:
```
title,id,parent_id,field_weight,field_identifier,field_resource_type,field_genre,field_linked_agent,field_model,field_edtf_date_created,field_place_published,field_description_long,field_subject,field_member_of,field_language,field_display_hints,file,field_domain_access
FEBRUARY 11 TEST COLLECTION,1,,,,collection,,,Collection,,,,,,,,,example_one_edu
Easthampton Town Hall (Large Image),2,1,,Easthampton Town Hall,still image,photograph,relators:cre:corporate_body:Historic American Buildings Survey|relators:pbl:corporate_body:Historic American Buildings Survey,Image,1933?,,Photo number 1. Photographs are albumen prints mounted on cardboard.,"subject:Black, Francis",,,Open Seadragon,EasthamptonTownHall.tif,example_one_edu
Nehemiah Strong House (Large Image),3,1,,Nehemiah Strong House,still image,photograph,relators:cre:corporate_body:Historic American Buildings Survey|relators:pbl:corporate_body:Historic American Buildings Survey,Image,1933?,,Photo number 2. Photographs are albumen prints mounted on cardboard.,"subject:Chittendon, Chester P.",,,Open Seadragon,NehemiahStrongHouse.tif,example_one_edu|example_two_edu
"Amherst College, Lawrence Observatory (Large Image)",4,1,,"Amherst College, Lawrence Observatory",still image,photograph,relators:cre:corporate_body:Historic American Buildings Survey|relators:pbl:corporate_body:Historic American Buildings Survey,Image,1933?,,Photo number 3. Photographs are albumen prints mounted on cardboard.,"subject:T'Shawn, David",,,Open Seadragon,AmherstCollegeLawrenceObservatory.tif,example_two_edu
```

**Comment by mjordan** [comment]
Fix merged with c1f6cd4a37163dc274e8142a7c120fa354d7791f, docs added to https://mjordan.github.io/islandora_workbench_docs/fields/#values-in-the-field_domain_access-column.
